tools/arcilator/arcilator.cpp

```diff
@@ -328,17 +336,48 @@ static LogicalResult processBuffer(
   if (!module)
     return failure();
 
-  PassManager pm(&context);
-  pm.enableVerifier(verifyPasses);
-  pm.enableTiming(ts);
-  if (failed(applyPassManagerCLOptions(pm)))
+  // Lower HwModule to Arc model.
+  PassManager pmArc(&context);
+  pmArc.enableVerifier(verifyPasses);
+  pmArc.enableTiming(ts);
+  if (failed(applyPassManagerCLOptions(pmArc)))
+    return failure();
+  populateHwModuleToArcPipeline(pmArc);
+
+  if (failed(pmArc.run(module.get())))
+    return failure();
+
+  // Collect generated model info.
+  std::vector<ModelInfo> models;
+  if (failed(collectModels(module.get(), models)))
+    return failure();
+
+  // Output state info as JSON if requested.
+  if (!stateFile.empty()) {
+    std::error_code ec;
+    llvm::ToolOutputFile outputFile(stateFile, ec,
+                                    llvm::sys::fs::OpenFlags::OF_None);
+    if (ec) {
+      llvm::errs() << "unable to open state file: " << ec.message() << '\n';
+      return failure();
+    }
+    serializeModelInfoToJson(outputFile.os(), models);
+    outputFile.keep();
```

Comment:
This should probably still have a place in the pipeline like a regular pass and not be executed when the pipeline ends earlier (as specified with `until-before` or `until-after`).

Reply:
Okay, I have achieved this by adding a check via untilReached to the emission of state file.